scrape/target.go

```diff
@@ -348,6 +338,95 @@ func (app *timeLimitAppender) Append(ref storage.SeriesRef, lset labels.Labels,
+// Use validateLabelSettings for sanity checks.
+func scrapeSettingsFromLabels(labels labels.Labels, defaults *labelSettings) (*labelSettings, error) {
+	var (
+		ls  = defaults
+		d   model.Duration
+		i64 int64
+		b2b units.Base2Bytes
+		err error
+	)
+
+	for _, l := range labels {
+		switch l.Name {
+		case model.ScrapeIntervalLabel:
+			if d, err = model.ParseDuration(l.Value); err == nil {
+				ls.interval = time.Duration(d)
+			}
+		case model.ScrapeTimeoutLabel:
+			if d, err = model.ParseDuration(l.Value); err == nil {
+				ls.timeout = time.Duration(d)
+			}
+		case model.SampleLimitLabel:
+			if i64, err = strconv.ParseInt(l.Value, 10, 0); err == nil {
+				ls.samleLimit = int(i64)
+			}
+		case model.BodySizeLimitLabel:
+			if err = b2b.UnmarshalText([]byte(l.Value)); err == nil {
+				ls.bodySizeLimit = int64(b2b)
+			}
+		case model.LabelLimitLabel:
+			if i64, err = strconv.ParseInt(l.Value, 10, 0); err == nil {
+				ls.labelLimits.labelLimit = int(i64)
+			}
+		case model.LabelNameLengthLimitLabel:
+			if i64, err = strconv.ParseInt(l.Value, 10, 0); err == nil {
+				ls.labelLimits.labelNameLengthLimit = int(i64)
+			}
+		case model.LabelValueLengthLimitLabel:
+			if i64, err = strconv.ParseInt(l.Value, 10, 0); err == nil {
+				ls.labelLimits.labelValueLengthLimit = int(i64)
+			}
+		}
+		if err != nil {
+			return ls, errors.Errorf("error parsing label %q with value %q: %v", l.Name, l.Value, err)
+		}
+	}
+	return ls, nil
+}
+
+// validateLabelSettings validates if settings are logically sane.
+func validateLabelSettings(ls *labelSettings) error {
+	if ls.interval == 0 {
+		return errors.New("scrape interval cannot be 0")
+	}
+	if ls.timeout == 0 {
+		return errors.New("scrape timeout cannot be 0")
+	}
+	if ls.timeout > ls.interval {
+		return errors.Errorf(
+			"scrape timeout cannot be greater than scrape interval (%q > %q)",
+			ls.timeout,
+			ls.interval,
+		)
+	}
+	if ls.samleLimit < 0 {
+		return errors.Errorf("sample limit cannot be negative (%q < 0)", ls.labelLimits)
+	}
+	if ls.bodySizeLimit < 0 {
+		return errors.Errorf("body size limit cannot be negative (%q < 0)", ls.bodySizeLimit)
+	}
+	if ls.labelLimits.labelLimit < 0 {
+		return errors.Errorf("label limit cannot be negative (%q < 0)", ls.labelLimits.labelLimit)
+	}
+	if ls.labelLimits.labelNameLengthLimit < 0 {
+		return errors.Errorf(
+			"label name length limit cannot be negative (%q < 0)",
+			ls.labelLimits.labelNameLengthLimit,
+		)
+	}
+	if ls.labelLimits.labelValueLengthLimit < 0 {
+		return errors.Errorf(
+			"label value length limit cannot be negative (%q < 0)",
+			ls.labelLimits.labelValueLengthLimit,
+		)
+	}
+	return nil
+}
+
 // PopulateLabels builds a label set from the given label set and scrape configuration.
 // It returns a label set before relabeling was applied as the second return value.
 // Returns the original discovered label set found before relabelling was applied if the target is dropped during relabeling.
```

Comment:
My belief is we don't need to set 
```
{Name: model.ScrapeIntervalLabel, Value: cfg.ScrapeInterval.String()},
{Name: model.ScrapeTimeoutLabel, Value: cfg.ScrapeTimeout.String()},
```
and other interval / timeout / honor labels, if they are not set already.
defaults are already there in the scrape config, there is no need to allocate strings if it could be avoided.

Reply:
They are allocated so that users can inspect relialably the configuration on each target. This is also what is documented. Chaging that is a breaking change.

Reply:
It being a breaking change is the reason those are left in place here.

A thought on configuration inspection: isn't it ultimately better to inspect the resulting typed values (`labelSettings`) instead of strings?

Reply:
> A thought on configuration inspection: isn't it ultimately better to inspect the resulting typed values (labelSettings) instead of strings?

I'm not sure I see the difference between inspecting the values from `labelSettings` and how we're currently doing it. Either way, they're going to be in the API response, just different fields, right? Target labels are the way that the values are set, and therefore users will expect them to be accessible from there, regardless if they've been modified.